torch/distributed/algorithms/join.py

```diff
@@ -65,55 +114,63 @@ class _Join():
     for details about the hook definition.
 
     .. warning::
-        The context manager requires a ``dist.all_reduce(torch.ones(1))`` to be
-        called on every non-joined process each time before it performs its
-        collective communications in order to indicate that the process has not
-        yet joined. For example, this can occur at the beginning of the forward
-        pass.
-
-    .. warning::
-        If ``throw_on_early_termination`` is enabled, then the context manager
-        additionally requires every non-joined process to participate in an
-        all-reduce before it performs its collective communications in order to
-        check if it should terminate due to detecting uneven inputs. This all-
-        reduce should be of the form ``dist.all_reduce(torch.zeros(1))``; if
-        the result is positive, then the process should terminate.
+        The context manager requires each participating :class:`_Joinable` to
+        call the method `notify_join_context()` before its own per-iteration
+        collective communications to ensure correctness.
 
     .. warning::
         The context manager requires that all ``process_group`` attributes in
         the ``_JoinHook`` objects are the same. If there are multiple
         ``_JoinHook`` objects, then the ``device`` of the first is used. The
         process group and device information is used for checking for non-
         joined processes and for notifying processes to terminate if
-        ``throw_on_early_termination`` is eanbled, both of which using an all-
+        ``throw_on_early_termination`` is enabled, both of which using an all-
         reduce.
 
     Arguments:
-        join_hooks (List[_JoinHook]): a list of the :class:`_JoinHook` s to
-            use; the hooks are iterated over in the given order.
+        joinables (List[_Joinable]): a list of the participating
+            :class:`_Joinable` s; their hooks are iterated over in the given
+            order.
 
         enable (bool): a flag enabling uneven input detection; setting to
             ``False`` disables the context manager's functionality and should
             only be set when the user knows the inputs will not be uneven
             (default: ``True``).
 
-        throw_on_early_termination (bool): a flag controlling whether to raise
-            an exception upon detecting uneven inputs (default: ``False``).
+        throw_on_early_termination (bool): a flag controlling whether to throw an
+            exception upon detecting uneven inputs (default: ``False``).
 
     """
     def __init__(
         self,
-        join_hooks: List[_JoinHook],
+        joinables: List[_Joinable],
         enable: bool = True,
         throw_on_early_termination: bool = False,
+        **kwargs,
```

Comment:
would I be correct that `is_first_joinable` should not be passed in kwargs?

Reply:
Correct. `is_first_joinable` is managed internally by the join context manager. The user never needs to interact with it.